v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -173,14 +160,14 @@ async def shutdown(self) -> None:
             self._keep_credentials_fresh_bg_task.cancel()
             cancelled_tasks.append(self._keep_credentials_fresh_bg_task)
 
-        # Wait for the cancellation to complete before returning
-        # NOTE: If cancelled while awaiting here, all tasks in gather will still be cancelled
-        # because the cancellations have already been issued.
-        # NOTE: Also, cancelling a gather implicitly cancels all the tasks that are gathered anyway
-        await asyncio.gather(*cancelled_tasks, return_exceptions=True)
-
-        if cached_exception:
-            raise cached_exception
+        results = await asyncio.gather(
```

Comment:
this can be done with asyncio.wait also right ? shield would not have been needed then.
any reason you chose this patten ?

Reply:
.wait only handles one awaitable at a time. It's also a lower level primitive, so I use a higher level abstraction when possible to benefit from higher level cancellation and error handling